core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -430,8 +519,54 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
     newIterator.foreach(addIfAbsent)
     oldIterator.foreach(addIfAbsent)
+    mergedApps
+  }
+
+
+  /**
+   * Scan through all incomplete applications, and check for any that have changed.
+   *
+   * After the scan, if there were any updated attempts, [[applications]] is updated
+   * with the new values.
+   *
+   * 1. No attempt to replay the application is made; this scan is a low cost operation.
+   * 2. As this overwrites [[applications]] with a new value, it must not run concurrently
+   */
+  private[history] def scanAndUpdateIncompleteAttemptInfo(): Unit = {
+    val now = System.currentTimeMillis();
+    val newAttempts: Iterable[FsApplicationAttemptInfo] = applications
+        .filter( e => !e._2.completed)
+        .flatMap { e =>
+          // build list of (false, attempt) or (true, attempt') values
+          e._2.attempts.flatMap { attempt: FsApplicationAttemptInfo =>
+            val path = new Path(logDir, attempt.logPath)
+            try {
+              val status = fs.getFileStatus(path)
+              val size = getLogSize(status).getOrElse(-1L)
+              val aS = attempt.fileSize
+              if (size > aS) {
```

Comment:
1. renamed
2. Not using timestamp, just size (and rm'd that unused System.getTimeMillis()) flag. There's some interesting race conditions which showed up in testing, which seemed fairly brittle to modtime changes. More succinctly: I don't think HDFS and Posix work the same w.r.t modtime.